examples_tests/0.ImportanceSamplingEnvMaps/main.cpp

```diff
@@ -0,0 +1,668 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#define _NBL_STATIC_LIB_
+#include <nabla.h>
+#include "../../../include/nbl/asset/filters/CSummedAreaTableImageFilter.h"
+#include "nbl/ext/FullScreenTriangle/FullScreenTriangle.h"
+#include "nbl/ext/ScreenShot/ScreenShot.h"
+#include "glm/glm.hpp"
+#include "../common/QToQuitEventReceiver.h"
+
+using namespace nbl;
+using namespace asset;
+using namespace core;
+using namespace video;
+
+using SATFilter = CSummedAreaTableImageFilter<false>;
+
+static core::smart_refctd_ptr<ICPUBuffer> computeLuminancePdf(smart_refctd_ptr<ICPUImage> envmap)
```

Comment:
I was expecting this to be implemented as a new IImageFilter (you could copy most of the impl from  a CopyAndConvert filter, which doesnt do any scaling, just decodes pixels from one image, saves to another, all you need to do is just add a modification step)

Reply:
actually upon thought I dont think you need and intermediate Luma image at all (could work directly from multi-channel SAT)

`Sum[dot(coeffs,rgb*sin(theta))] == dot(coeffs,Sum[rgb*sin(theta)])`

Also upon further reflection, a slight modification to `CSwizzleableAndDitherableFilterBase` (refactor into a `TranscodeFilterBase`) to allow a generalpurpose `encodePrelude` and `decodePostlude` lambdas to be called in the first thing in `onEncode` and last thing in `onDecode` (and have the `CSwizzleableAndDitherableFilterBase` refactor its `onEncode` and `onDecode` into preludes and postludes that apply swizzles and dithers) would enable the implementation `sin(theta)` weighting (and luma computation if you wish) as part of an image filter.

Then you could do
```cpp
using EnvmapLuminancePdfComputationFilter = TranscodeFilterBase<decltype(lumaDotLambda),decltype(sinThetaLamda)>; // cvt to luma on decode, weigh by sin on encode
```

Reply:
Or, even less effort, you could add the custom per-pixel lambda to the SAT filter instead!

change signature to 
```cpp
template<bool ExclusiveMode = false, class CustomPerPixelLambda=CSummedAreaTableImageFilterBase::IdentityPerPixelLambda/*NOOP*/>
class CSummedAreaTableImageFilter 
```
and add a `CustomPerPixelLambda*` to the `CState` if it inherits from `CSummedAreaTableImageFilterBase::StatefulPerPixelLambda`

then right after the any calls to `asset::decodePixelsRuntime` you can do 
```cpp
if constexpr (std::is_base_of_v<CSummedAreaTableImageFilterBase::StatefulPerPixelLambda,CustomPerPixelLambda>)
    state.perPixelLambda(decodeBuffer,blockX,blockY);
else
    CustomPerPixelLambda()(decodeBuffer,blockX,blockY);
```